File: legal-automation/backend/app/schemas/ki.py
from datetime import datetime

from pydantic import BaseModel, field_validator

from app.models.legal_knowledge import SOURCE_TYPES


class KiQueryRequest(BaseModel):
    question: str
    matter_id: int | None = None

    @field_validator("question")
    @classmethod
    def question_not_empty(cls, v: str) -> str:
        if len(v.strip()) < 5:
            raise ValueError("Frage zu kurz")
        return v.strip()


class KiSource(BaseModel):
    marker: str
    chunk_id: int
    heading: str | None
    document_title: str
    source_type: str
    external_id: str | None
    url_or_ref: str | None


class KiQueryResponse(BaseModel):
    query_id: int | None
    answer: str
    grounded: bool
    sources: list[KiSource]
    model: str
    disclaimer: str = "KI-Entwurf — keine Rechtsberatung, anwaltliche Prüfung erforderlich."


class KiIngestRequest(BaseModel):
    source_type: str
    title: str
    text: str
    external_id: str | None = None
    jurisdiction: str | None = None
    url_or_ref: str | None = None
    matter_id: int | None = None

    @field_validator("source_type")
    @classmethod
    def valid_source_type(cls, v: str) -> str:
        if v not in SOURCE_TYPES:
            raise ValueError(f"source_type must be one of {sorted(SOURCE_TYPES)}")
        return v

    @field_validator("text")
    @classmethod
    def text_not_empty(cls, v: str) -> str:
        if len(v.strip()) < 20:
            raise ValueError("Dokumenttext zu kurz")
        return v


class KiIngestResponse(BaseModel):
    document_id: int | None
    num_chunks: int
    duplicate: bool


class KiDocumentResponse(BaseModel):
    id: int
    source_type: str
    external_id: str | None
    title: str
    jurisdiction: str | None
    matter_id: int | None
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class KiFeedbackRequest(BaseModel):
    feedback: str  # up | down
    note: str | None = None

    @field_validator("feedback")
    @classmethod
    def valid_feedback(cls, v: str) -> str:
        if v not in {"up", "down"}:
            raise ValueError("feedback must be 'up' or 'down'")
        return v


class KiStatusResponse(BaseModel):
    enabled: bool
    ollama_available: bool
    llm_model: str
    embed_model: str
    num_documents: int
    num_chunks: int

File: legal-automation/frontend/src/lib/api/ki.ts
import api from "./client";

export interface KiSource {
  marker: string;
  chunk_id: number;
  heading: string | null;
  document_title: string;
  source_type: string;
  external_id: string | null;
  url_or_ref: string | null;
}

export interface KiQueryResult {
  query_id: number | null;
  answer: string;
  grounded: boolean;
  sources: KiSource[];
  model: string;
  disclaimer: string;
}

export interface KiStatus {
  enabled: boolean;
  ollama_available: boolean;
  llm_model: string;
  embed_model: string;
  num_documents: number;
  num_chunks: number;
}

export interface KiDocument {
  id: number;
  source_type: string;
  external_id: string | null;
  title: string;
  jurisdiction: string | null;
  matter_id: number | null;
  is_active: boolean;
  created_at: string;
}

export const kiApi = {
  status: () => api.get<KiStatus>("/ki/status").then((r) => r.data),

  query: (question: string, matter_id?: number) =>
    api.post<KiQueryResult>("/ki/query", { question, matter_id }).then((r) => r.data),

  ingest: (data: {
    source_type: string;
    title: string;
    text: string;
    external_id?: string;
    matter_id?: number;
  }) =>
    api
      .post<{ document_id: number | null; num_chunks: number; duplicate: boolean }>(
        "/ki/ingest",
        data
      )
      .then((r) => r.data),

  documents: () => api.get<KiDocument[]>("/ki/documents").then((r) => r.data),

  feedback: (queryId: number, feedback: "up" | "down", note?: string) =>
    api.post(`/ki/queries/${queryId}/feedback`, { feedback, note }).then((r) => r.data),
};
